### ISSUE TITLE
double free or corruption (!prev)

### ISSUE THREAD

**Issue by VisualPrankDude** [radare/radare2#2196]
ok. this is hardcore. just tried to analyse all 17000 png files on my GNU/Linux system:
 
```locate -0 \.png | xargs --verbose --null  r2 -A -c "i~uri" -d  ```

valgrind Output:

```
[...] 
/zenity/clothes/monk.png /usr/share/zenity/clothes/sunglasses.png
 /usr/share/zenity/clothes/surfboard.png
==9298== Parent PID: 9194
==9298== 
==9301== Invalid write of size 1
==9301==    at 0x4DF4AA3: r_str_argv (str.c:1159)
==9301==    by 0x47376BC: fork_and_ptraceme (io_debug.c:271)
==9301==    by 0x47378C7: __open (io_debug.c:363)
==9301==    by 0x474042E: __getioplugin (io.c:111)
==9301==    by 0x47406F1: r_io_open_nomap (io.c:185)
==9301==    by 0x40A591F: r_core_file_open (file.c:628)
==9301==    by 0x804B9E9: main (radare2.c:521)
==9301==  Address 0x51f63eb is 0 bytes after a block of size 39,043 alloc'd
==9301==    at 0x4028308: malloc (vg_replace_malloc.c:263)
==9301==    by 0x4DF48A5: r_str_argv (str.c:1080)
==9301==    by 0x47376BC: fork_and_ptraceme (io_debug.c:271)
==9301==    by 0x47378C7: __open (io_debug.c:363)
==9301==    by 0x474042E: __getioplugin (io.c:111)
==9301==    by 0x47406F1: r_io_open_nomap (io.c:185)
==9301==    by 0x40A591F: r_core_file_open (file.c:628)
==9301==    by 0x804B9E9: main (radare2.c:521)
==9301== 
==9301== Invalid read of size 1
==9301==    at 0x4028A73: __GI_strlen (mc_replace_strmem.c:391)
==9301==    by 0x4EC4EB4: strdup (strdup.c:42)
==9301==    by 0x4DF4AC9: r_str_argv (str.c:1160)
==9301==    by 0x47376BC: fork_and_ptraceme (io_debug.c:271)
==9301==    by 0x47378C7: __open (io_debug.c:363)
==9301==    by 0x474042E: __getioplugin (io.c:111)
==9301==    by 0x47406F1: r_io_open_nomap (io.c:185)
==9301==    by 0x40A591F: r_core_file_open (file.c:628)
==9301==    by 0x804B9E9: main (radare2.c:521)
==9301==  Address 0x51f63eb is 0 bytes after a block of size 39,043 alloc'd
==9301==    at 0x4028308: malloc (vg_replace_malloc.c:263)
==9301==    by 0x4DF48A5: r_str_argv (str.c:1080)
==9301==    by 0x47376BC: fork_and_ptraceme (io_debug.c:271)
==9301==    by 0x47378C7: __open (io_debug.c:363)
==9301==    by 0x474042E: __getioplugin (io.c:111)
==9301==    by 0x47406F1: r_io_open_nomap (io.c:185)
==9301==    by 0x40A591F: r_core_file_open (file.c:628)
==9301==    by 0x804B9E9: main (radare2.c:521)
==9301== 
==9301== Invalid read of size 4
==9301==    at 0x4EC6966: __GI_memcpy (memcpy.S:75)
==9301==    by 0x4DF4AC9: r_str_argv (str.c:1160)
==9301==    by 0x47376BC: fork_and_ptraceme (io_debug.c:271)
==9301==    by 0x47378C7: __open (io_debug.c:363)
==9301==    by 0x474042E: __getioplugin (io.c:111)
==9301==    by 0x47406F1: r_io_open_nomap (io.c:185)
==9301==    by 0x40A591F: r_core_file_open (file.c:628)
==9301==    by 0x804B9E9: main (radare2.c:521)
==9301==  Address 0x51f63e8 is 39,040 bytes inside a block of size 39,043 alloc'd
==9301==    at 0x4028308: malloc (vg_replace_malloc.c:263)
==9301==    by 0x4DF48A5: r_str_argv (str.c:1080)
==9301==    by 0x47376BC: fork_and_ptraceme (io_debug.c:271)
==9301==    by 0x47378C7: __open (io_debug.c:363)
==9301==    by 0x474042E: __getioplugin (io.c:111)
==9301==    by 0x47406F1: r_io_open_nomap (io.c:185)
==9301==    by 0x40A591F: r_core_file_open (file.c:628)
==9301==    by 0x804B9E9: main (radare2.c:521)
==9301== 
==9301== 
==9301== HEAP SUMMARY:
==9301==     in use at exit: 514,533 bytes in 3,439 blocks
==9301==   total heap usage: 7,424 allocs, 3,985 frees, 25,193,441 bytes allocated
==9301== 
==9301== LEAK SUMMARY:
==9301==    definitely lost: 384 bytes in 23 blocks
==9301==    indirectly lost: 2,743 bytes in 48 blocks
==9301==      possibly lost: 11,057 bytes in 39 blocks
==9301==    still reachable: 500,349 bytes in 3,329 blocks
==9301==         suppressed: 0 bytes in 0 blocks
==9301== Rerun with --leak-check=full to see details of leaked memory
==9301== 
==9301== For counts of detected and suppressed errors, rerun with: -v
==9301== ERROR SUMMARY: 3 errors from 3 contexts (suppressed: 721 from 12)
==9298== 
==9298== HEAP SUMMARY:
==9298==     in use at exit: 433,887 bytes in 2,800 blocks
==9298==   total heap usage: 6,785 allocs, 3,985 frees, 25,073,939 bytes allocated
==9298== 
==9298== LEAK SUMMARY:
==9298==    definitely lost: 408 bytes in 25 blocks
==9298==    indirectly lost: 2,755 bytes in 49 blocks
==9298==      possibly lost: 50,731 bytes in 41 blocks
==9298==    still reachable: 379,993 bytes in 2,685 blocks
==9298==         suppressed: 0 bytes in 0 blocks
==9298== Rerun with --leak-check=full to see details of leaked memory
==9298== 
==9298== For counts of detected and suppressed errors, rerun with: -v
==9298== ERROR SUMMARY: 0 errors from 0 contexts (suppressed: 721 from 12)
```

**Comment by VisualPrankDude** [comment]
This is one example:

```
r2 -A -c "i~uri" -d \(EgotisticalGiraffe\)
*** glibc detected *** r2: malloc(): memory corruption: 0x08a8c5f0 ***
======= Backtrace: =========
/lib/i386-linux-gnu/i686/cmov/libc.so.6(+0x6feb1)[0xb682feb1]
/lib/i386-linux-gnu/i686/cmov/libc.so.6(+0x72c94)[0xb6832c94]
/lib/i386-linux-gnu/i686/cmov/libc.so.6(__libc_malloc+0x5c)[0xb683492c]
/lib/i386-linux-gnu/i686/cmov/libc.so.6(__strdup+0x30)[0xb6837ec0]
/usr/local/lib/libr_util.so(r_str_argv+0x28e)[0xb6947aca]
/usr/local/lib/libr_io.so(+0xa6bd)[0xb70006bd]
/usr/local/lib/libr_io.so(+0xa8c8)[0xb70008c8]
/usr/local/lib/libr_io.so(+0x1342f)[0xb700942f]
/usr/local/lib/libr_io.so(r_io_open_nomap+0x46)[0xb70096f2]
/usr/local/lib/libr_core.so(r_core_file_open+0xc5)[0xb76b2920]
```

**Comment by nero-luci** [comment]
I am not sure, but I suppose this bug can be used for a heap-based overflow. So it may be critical :-?
Can you help with some other details (maybe some pointers to the code, because I am new to the codebase). Thank you!

**Comment by radare** [comment]
Yep. All crashes are critical, join the irc for discussion

**Comment by zonkzonk** [comment]
still in 285b20ce1ad124d9a7c32220572b73f5e76aaefe

**Comment by nero-luci** [comment]
@zonkzonk https://github.com/radare/radare2/commit/285b20ce1ad124d9a7c32220572b73f5e76aaefe solves this bug?

**Comment by zonkzonk** [comment]
@nero-luci  nope

**Comment by radare** [comment]
Can you confirm again with latest git?

**Comment by VisualPrankDude** [comment]
cool, this one is fixed in 261a4d2faa9e7004f01fae17ca44036f16c6467b :)
